### PR TITLE
feat: add submit funding proposal

### PR DESCRIPTION
**Pull request by 0xChin**
Closes OPT-785

  ## 🚨 Spec-Breaking Changes

  This implementation includes two significant deviations from the original specification that require discussion and approval:

  ### 1. Changed `proposalVotingModule` Type: `uint8` → `address`

  **Original Spec**: `ProposalTypeData.proposalVotingModule` was defined as `uint8`
  **Implementation**: Changed to `address`

  **Justification**:
  - We are not integrating with the OptimismGovernor's proposal type system (which uses uint8 IDs)
  - Our governance flow is independent and uses its own proposal type management
  - Direct address reference eliminates unnecessary indirection and is more gas-efficient
  - Provides better type safety and clarity in our architecture

  **Impact**: All structs, interfaces, and function signatures now use `address` instead of `uint8`

  ### 2. Interpreted `criteriaValue` as Absolute Token Count (Not Percentage)

  **Original Spec/NatSpec**: Described `criteriaValue` as "percentage that will be used to calculate the fraction of the votable supply"
  **Implementation**: Treats `criteriaValue` as absolute token count required for options to pass

  **Justification**:
  - ApprovalVotingModule code clearly [expects absolute token amounts](https://github.com/defi-wonderland/optimism/blob/b169a6cbc10119ac86a2a03cbcdb0dbc42d544eb/packages/contracts-bedrock/src/governance/ApprovalVotingModule.sol#L346-L349) (direct comparison: `optionVotes[i] >= settings.criteriaValue`) 
  - No percentage conversion logic exists anywhere in the codebase
  - Voting weights are passed as raw token amounts, not percentages
  - Absolute token thresholds are more predictable and commonly used in governance systems
  - The original NatSpec appears to be misleading/incorrect based on actual implementation

  **Impact**: Functions expect token amounts (e.g., `1000 ether`) rather than percentage values

  ---

  **Note**: These changes align the implementation with the actual ApprovalVotingModule contract behavior and our independent governance architecture.

### CODE REVIEW

File: packages/contracts-bedrock/src/governance/ProposalValidator.sol [hunk 1]
```diff
@@ -227,6 +251,98 @@ contract ProposalValidator is OwnableUpgradeable, ReinitializableBase, ISemver {
...
+    /// that will be used to calculate the fraction of the votable supply that the proposal will need in votes in order
+    /// to pass.
+    /// @param _optionsDescriptions The strings of the different options that can be voted.
+    /// @param _optionsRecipients An address for each option to transfer funds to in case the option passes the voting.
+    /// @param _optionsAmounts The amount to transfer for each option in case the option passes the voting.
+    /// @param _description Description of the proposal.
+    /// @param _proposalType The type of proposal (must be GovernanceFund or CouncilBudget).
+    /// @return proposalHash_ The hash of the submitted proposal.
+    function submitFundingProposal(
+        uint128 _criteriaValue,
```
Comment (0xOneTony) [inline packages/contracts-bedrock/src/governance/ProposalValidator.sol:278]: Thought about this a bit more, I believe we should stick with the specs, the reason we use the percentage here is because since this function is permissionless we want the user to have an easier experience on providing these data. The idea is that they provide the percentage and we use that to calculate the amount of tokens **relative** to the existing voting supply

File: packages/contracts-bedrock/src/governance/ProposalValidator.sol [hunk 2]
```diff
@@ -227,6 +251,98 @@ contract ProposalValidator is OwnableUpgradeable, ReinitializableBase, ISemver {
...
+        string[] memory _optionsDescriptions,
+        address[] memory _optionsRecipients,
+        uint256[] memory _optionsAmounts,
+        string memory _description,
+        ProposalType _proposalType
+    )
+        external
+        returns (bytes32 proposalHash_)
+    {
+        if (_proposalType != ProposalType.GovernanceFund && _proposalType != ProposalType.CouncilBudget) {
```
Comment (0xOneTony) [inline packages/contracts-bedrock/src/governance/ProposalValidator.sol:289]: I would like to see more in-line comments for every step

**Review by 0xOneTony — CHANGES_REQUESTED** [review]
Also still thinking about storing **only** the address instead of the voting module type. Afaik the address of the voting module can change for a specific type so this becomes an issue if we dont have also a function to change that address. I will get back to this.

File: packages/contracts-bedrock/src/governance/ProposalValidator.sol (hunk 1 above)
Comment (0xChin) [reply packages/contracts-bedrock/src/governance/ProposalValidator.sol:278]: I agree with making the experience better. But would like to further discuss this, I have this key points:

  1. **UX should be handled in the UI** - The frontend can easily calculate percentages and convert them to absolute values before calling the contract
  2. **Adding percentage logic creates inconsistency** - It would make the 1submitFundingProposal1 function behave differently from how 1ApprovalVotingModule1 expects 1criteriaValue1 to work
  3. **If percentages were implemented**, the parameter should indeed be named 1criteriaValuePercentage1 to make the intent clear

File: packages/contracts-bedrock/src/governance/ProposalValidator.sol (hunk 2 above)
Comment (0xChin) [reply packages/contracts-bedrock/src/governance/ProposalValidator.sol:289]: [chore: add comments for submitFundingProposal](https://github.com/defi-wonderland/optimism/commit/5654e5699)

File: packages/contracts-bedrock/src/governance/ProposalValidator.sol [hunk 3]
```diff
@@ -227,6 +251,106 @@ contract ProposalValidator is OwnableUpgradeable, ReinitializableBase, ISemver {
...
+        bytes memory proposalVotingModuleData = abi.encode(options, settings);
+
+        // Generate unique proposal hash
+        proposalHash_ = _hashProposalWithModule(
+            proposalTypesData[_proposalType].proposalVotingModule,
+            proposalVotingModuleData,
+            keccak256(bytes(_description))
+        );
+
+        ProposalData storage proposal = _proposals[proposalHash_];
```
Comment (0xOneTony) [inline packages/contracts-bedrock/src/governance/ProposalValidator.sol:349]: We need to make sure also that this proposal does not exist on the Governor aswell, so we will need a check like `governor.proposalDeadline(proposalHash_) != 0`

File: packages/contracts-bedrock/src/governance/ProposalValidator.sol [hunk 4]
```diff
@@ -71,7 +81,7 @@ contract ProposalValidator is OwnableUpgradeable, ReinitializableBase, ISemver {
     /// @param proposalVotingModule The voting module each proposal type must use.
     struct ProposalTypeData {
         uint256 requiredApprovals;
-        uint8 proposalVotingModule;
+        address proposalVotingModule;
```
Comment (0xOneTony) [inline packages/contracts-bedrock/src/governance/ProposalValidator.sol]: After thinking about this a bit more, I think we should keep both the address and the uin8 id. Eventually we will be calling `governor.proposeWithModule` which need both of these arguments. The things is we cant have a good way of getting this id other than the on initiallize or manager set this value.

Also I see an improvement on the way we set the voting module address. Currently you expect the owner to have this address and set it on initiallize or with the setter. However we should use the `uint8` id and the `ProposalTypeConfigurator` contract to get the module address. In the future in case a new voting module is introduced i dont want OPF to have to set the same address in both the type configurator and the proposal validator contract.

File: packages/contracts-bedrock/src/governance/ProposalValidator.sol (hunk 1 above)
Comment (0xOneTony) [reply packages/contracts-bedrock/src/governance/ProposalValidator.sol:278]: The second point does not matter since the idea is that we calculate in the code the amount of votes and we provide that to the voting module.
The third point also does not add any value on keeping this as a pure number.
I hear your first point though and this can reduce some complexity for the contract. Lets keep it as is for now with the voting number instead of the percentage but we should def change the natspec for the param and add a dev comment specific for this explaining the UX and what the user who would call this function from the UI vs someone who calls this directly would need to know.

File: packages/contracts-bedrock/src/governance/ProposalValidator.sol [hunk 5]
```diff
@@ -227,6 +251,106 @@ contract ProposalValidator is OwnableUpgradeable, ReinitializableBase, ISemver {
...
+        external
+        returns (bytes32 proposalHash_)
+    {
+        // Only funding proposal types can use this function
+        if (_proposalType != ProposalType.GovernanceFund && _proposalType != ProposalType.CouncilBudget) {
+            revert ProposalValidator_InvalidFundingProposalType();
+        }
+
+        // Validate input arrays have matching lengths
+        uint256 optionsLength = _optionsDescriptions.length;
```
Comment (0xOneTony) [inline packages/contracts-bedrock/src/governance/ProposalValidator.sol:294]: regarding the options we should also have this check
`if (optionsLength == 0 || optionsLength > type(uint8).max) revert`

**Review by 0xOneTony — CHANGES_REQUESTED** [review]

File: packages/contracts-bedrock/src/governance/ProposalValidator.sol (hunk 1 above)
Comment (0xChin) [reply packages/contracts-bedrock/src/governance/ProposalValidator.sol:278]: [docs: improve natspec](https://github.com/defi-wonderland/optimism/commit/6d5a83b05)

File: packages/contracts-bedrock/src/governance/ProposalValidator.sol (hunk 5 above)
Comment (0xChin) [reply packages/contracts-bedrock/src/governance/ProposalValidator.sol:294]: [feat: add validation for options length in ProposalValidator](https://github.com/defi-wonderland/optimism/commit/f599dc16a)

File: packages/contracts-bedrock/src/governance/ProposalValidator.sol (hunk 3 above)
Comment (0xChin) [reply packages/contracts-bedrock/src/governance/ProposalValidator.sol:349]: There already exists a check for this in the governor [implementation](https://github.com/voteagora/optimism-governor/blob/3f8a2ea29d2c91a176758c72a213d499424ee2fc/src/OptimismGovernor.sol#L449-L450), [test](https://github.com/voteagora/optimism-governor/blob/3f8a2ea29d2c91a176758c72a213d499424ee2fc/test/OptimismGovernor.t.sol#L384-L398)

On the other hand, the hash created from the `hashProposalWithModule` is different in each contract so we would need to call `governor.hashProposalWithModule(...)`. The reason of this is that one of the args for constructing the hash is the `address(this)`, which is different in each contract obviously. Should we construct the hash in the validator based on the governor's address?

File: packages/contracts-bedrock/src/governance/ProposalValidator.sol (hunk 4 above)
Comment (0xChin) [reply packages/contracts-bedrock/src/governance/ProposalValidator.sol]: [feat: get votingModuleAddress from ProposalTypeConfigurator](https://github.com/defi-wonderland/optimism/commit/c40754bdd)

File: packages/contracts-bedrock/src/governance/ProposalValidator.sol (hunk 3 above)
Comment (0xOneTony) [reply packages/contracts-bedrock/src/governance/ProposalValidator.sol:349]: We cant have a "passing" proposal from the validator failling on the governor. We should have the check in both places. 
Like mentioned in the past, we should produce the SAME proposal hash with the governor, this means if our hash function uses `address(this)` it is wrong, and it should be `address(governor)`

File: packages/contracts-bedrock/src/governance/ProposalValidator.sol (hunk 5 above)
Comment (0xOneTony) [reply packages/contracts-bedrock/src/governance/ProposalValidator.sol:294]: No tests?

File: packages/contracts-bedrock/src/governance/ProposalValidator.sol (hunk 3 above)
Comment (0xChin) [reply packages/contracts-bedrock/src/governance/ProposalValidator.sol:349]: [feat: check for proposal existance on submission](https://github.com/defi-wonderland/optimism/commit/088748bf2)

Comment (0xChin) [reply packages/contracts-bedrock/src/governance/ProposalValidator.sol:349]: Given this new integration, we could maybe remove the `inVoting` field from the `ProposalData` struct and just make the external call. I'm not adding it given the `moveToVote` function development is on hold

File: packages/contracts-bedrock/src/governance/ProposalValidator.sol (hunk 5 above)
Comment (0xChin) [reply packages/contracts-bedrock/src/governance/ProposalValidator.sol:294]: [test: add InvalidOptionsLength tests](https://github.com/defi-wonderland/optimism/commit/508c3c68a)

File: packages/contracts-bedrock/test/governance/ProposalValidator.t.sol [hunk 6]
```diff
@@ -578,11 +656,583 @@ contract ProposalValidator_HashProposalWithModule_Test is ProposalValidator_Init
...
+            criteriaValue,
+            optionsDescriptions,
+            optionsRecipients,
+            optionsAmounts,
+            description,
+            ProposalValidator.ProposalType.GovernanceFund
+        );
+    }
+
+    function test_submitFundingProposal_duplicateProposal_reverts() public {
```
Comment (0xOneTony) [inline packages/contracts-bedrock/test/governance/ProposalValidator.t.sol:955]: Some of these tests have the same name or even logic

File: packages/contracts-bedrock/test/governance/ProposalValidator.t.sol [hunk 7]
```diff
@@ -578,11 +656,583 @@ contract ProposalValidator_HashProposalWithModule_Test is ProposalValidator_Init
...
+            criteriaValue,
+            optionsDescriptions,
+            optionsRecipients,
+            optionsAmounts,
+            description,
+            ProposalValidator.ProposalType.GovernanceFund
+        );
+    }
+
+    function test_submitFundingProposal_multipleAmountsExceedThreshold_reverts() public {
```
Comment (0xOneTony) [inline packages/contracts-bedrock/test/governance/ProposalValidator.t.sol]: this

File: packages/contracts-bedrock/test/governance/ProposalValidator.t.sol [hunk 8]
```diff
@@ -578,11 +656,583 @@ contract ProposalValidator_HashProposalWithModule_Test is ProposalValidator_Init
...
```
Comment (0xOneTony) [inline packages/contracts-bedrock/test/governance/ProposalValidator.t.sol]: and this

**Review by 0xOneTony — CHANGES_REQUESTED** [review]
There are some names and tests that are the same lets check all of them to make sure we dont repeat code

File: packages/contracts-bedrock/test/governance/ProposalValidator.t.sol (hunk 6 above)
Comment (0xChin) [reply packages/contracts-bedrock/test/governance/ProposalValidator.t.sol:955]: [fix: remove duplicated tests](https://github.com/defi-wonderland/optimism/commit/a3c9a9465)

File: packages/contracts-bedrock/test/governance/ProposalValidator.t.sol (hunk 7 above)
Comment (0xChin) [reply packages/contracts-bedrock/test/governance/ProposalValidator.t.sol]: [fix: remove duplicated tests](https://github.com/defi-wonderland/optimism/commit/a3c9a9465)

File: packages/contracts-bedrock/test/governance/ProposalValidator.t.sol (hunk 8 above)
Comment (0xChin) [reply packages/contracts-bedrock/test/governance/ProposalValidator.t.sol]: [fix: remove duplicated tests](https://github.com/defi-wonderland/optimism/commit/a3c9a9465)

File: packages/contracts-bedrock/src/governance/ProposalValidator.sol [hunk 9]
```diff
@@ -227,6 +261,118 @@ contract ProposalValidator is OwnableUpgradeable, ReinitializableBase, ISemver {
...
+            revert ProposalValidator_ProposalTypesDataLengthMismatch();
+        }
+
+        // Validate options length bounds
+        if (optionsLength == 0 || optionsLength > type(uint8).max) {
+            revert ProposalValidator_InvalidOptionsLength();
+        }
+
+        // Check each option amount against distribution threshold
+        for (uint256 i = 0; i < optionsLength; i++) {
```
Comment (0xOneTony) [inline packages/contracts-bedrock/src/governance/ProposalValidator.sol:308]: All 3 `for` loops have the same length, can we merge them into 1 for loop? the minimum would be to at least merge the building with the total budget, but i think it makes sense to merge all 3into 1

File: packages/contracts-bedrock/test/governance/ProposalValidator.t.sol [hunk 10]
```diff
@@ -578,11 +656,497 @@ contract ProposalValidator_HashProposalWithModule_Test is ProposalValidator_Init
...
+        // Mock proposalSnapshot to return 0 (proposal doesn't exist in governor)
+        _mockAndExpect(
+            address(governor),
+            abi.encodeCall(IOptimismGovernor.proposalSnapshot, (uint256(expectedHash))),
+            abi.encode(0)
+        );
+
+        // Expect ProposalSubmitted event
+        vm.expectEmit(address(validator));
+        emit ProposalSubmitted(expectedHash, rando, description, ProposalValidator.ProposalType.GovernanceFund);
```
Comment (0xOneTony) [inline packages/contracts-bedrock/test/governance/ProposalValidator.t.sol]: I think it would be better if we fuzzed the proposer instead of using a "random" address

File: packages/contracts-bedrock/test/governance/ProposalValidator.t.sol [hunk 11]
```diff
@@ -578,11 +656,497 @@ contract ProposalValidator_HashProposalWithModule_Test is ProposalValidator_Init
...
+        bytes32 proposalHash = validator.submitFundingProposal(
+            criteriaValue,
+            optionsDescriptions,
+            optionsRecipients,
+            optionsAmounts,
+            description,
+            ProposalValidator.ProposalType.GovernanceFund
+        );
+
+        assertEq(proposalHash, expectedHash);
```
Comment (0xOneTony) [inline packages/contracts-bedrock/test/governance/ProposalValidator.t.sol]: I think this is not enough, we should at least also check that the proposal was stored with the correct data

File: packages/contracts-bedrock/test/governance/ProposalValidator.t.sol [hunk 12]
```diff
@@ -578,11 +656,497 @@ contract ProposalValidator_HashProposalWithModule_Test is ProposalValidator_Init
...
+    }
+
+    function test_submitFundingProposal_councilBudget_succeeds() public {
```
Comment (0xOneTony) [inline packages/contracts-bedrock/test/governance/ProposalValidator.t.sol]: Same comments on this and the rest of the tests in the scope of this PR that do the same

File: packages/contracts-bedrock/test/governance/ProposalValidator.t.sol [hunk 13]
```diff
@@ -578,11 +656,497 @@ contract ProposalValidator_HashProposalWithModule_Test is ProposalValidator_Init
...
+        singleRecipient[0] = makeAddr("singleRecipient");
+
+        uint256[] memory singleAmount = new uint256[](1);
+        singleAmount[0] = 100 ether;
+
+        // Calculate expected proposal hash
+        bytes memory singleOptionData =
+            _constructVotingModuleData(singleDescription, singleRecipient, singleAmount, criteriaValue);
+        bytes32 expectedHash = validator.hashProposalWithModule(
+            makeAddr("approvalVotingModule"), singleOptionData, keccak256(bytes(description))
```
Comment (0xOneTony) [inline packages/contracts-bedrock/test/governance/ProposalValidator.t.sol]: we create this address again and again, why we dont define it in the contract once?

File: packages/contracts-bedrock/test/governance/ProposalValidator.t.sol [hunk 14]
```diff
@@ -578,11 +656,497 @@ contract ProposalValidator_HashProposalWithModule_Test is ProposalValidator_Init
...
+            description,
+            ProposalValidator.ProposalType.CouncilBudget
+        );
+
+        assertEq(proposalHash, expectedHash);
+    }
+
+    function test_submitFundingProposal_singleOption_succeeds() public {
+        string[] memory singleDescription = new string[](1);
+        singleDescription[0] = "Single option";
```
Comment (0xOneTony) [inline packages/contracts-bedrock/test/governance/ProposalValidator.t.sol]: I think it makes more sense to fuzz these values

File: packages/contracts-bedrock/test/governance/ProposalValidator.t.sol [hunk 15]
```diff
@@ -578,11 +656,497 @@ contract ProposalValidator_HashProposalWithModule_Test is ProposalValidator_Init
...
+        bytes32 proposalHash = validator.submitFundingProposal(
+            criteriaValue,
+            singleDescription,
+            singleRecipient,
+            singleAmount,
+            description,
+            ProposalValidator.ProposalType.GovernanceFund
+        );
+
+        assertTrue(proposalHash != bytes32(0));
```
Comment (0xOneTony) [inline packages/contracts-bedrock/test/governance/ProposalValidator.t.sol]: Why not use the expected hash in this case?

File: packages/contracts-bedrock/test/governance/ProposalValidator.t.sol [hunk 16]
```diff
@@ -578,11 +656,497 @@ contract ProposalValidator_HashProposalWithModule_Test is ProposalValidator_Init
...
+            optionsRecipients,
+            optionsAmounts,
+            description,
+            ProposalValidator.ProposalType.CouncilBudget
+        );
+
+        assertEq(proposalHash, expectedHash);
+    }
+
+    function test_submitFundingProposal_singleOption_succeeds() public {
```
Comment (0xOneTony) [inline packages/contracts-bedrock/test/governance/ProposalValidator.t.sol]: some of the first tests where called twice for both proposal types, the rest of the tests only cover the `Governance Fund` proposals only. I think we could fuzz between the 2 proposal types instead of having 2 times the same test or only test 1 for 1 of the types

Comment (0xOneTony) [reply packages/contracts-bedrock/test/governance/ProposalValidator.t.sol]: Same with criteriaValue, we could accept a range of accepted values for criteria value instead of using a single hardcoded one

Comment (0xOneTony) [reply packages/contracts-bedrock/test/governance/ProposalValidator.t.sol]: testing edge cases like `test_submitFundingProposal_maximumThreshold_succeeds` is good if you dont fuzz the specific input since it affects the edge case, but the rest of the values that makes sense to be fuzzed should be fuzzed in their acceptable range of the happy path

File: packages/contracts-bedrock/test/governance/ProposalValidator.t.sol [hunk 17]
```diff
@@ -578,11 +656,497 @@ contract ProposalValidator_HashProposalWithModule_Test is ProposalValidator_Init
...
+
+        optionsRecipients = new address[](2);
+        optionsRecipients[0] = makeAddr("recipient1");
+        optionsRecipients[1] = makeAddr("recipient2");
+
+        optionsAmounts = new uint256[](2);
+        optionsAmounts[0] = 1000 ether;
+        optionsAmounts[1] = 500 ether;
+
+        description = "Test funding proposal";
```
Comment (0xOneTony) [inline packages/contracts-bedrock/test/governance/ProposalValidator.t.sol]: This setup is repeated in `ProposalValidator_SubmitFundingProposal_Test` aswell, maybe defining the proposal type data is something that could be added in the `Init` contract since this will probably be needed to happen in most of the tests?

File: packages/contracts-bedrock/test/governance/ProposalValidator.t.sol [hunk 18]
```diff
@@ -578,11 +656,497 @@ contract ProposalValidator_HashProposalWithModule_Test is ProposalValidator_Init
...
+    function test_submitFundingProposal_invalidProposalType_reverts() public {
+        vm.expectRevert(ProposalValidator.ProposalValidator_InvalidFundingProposalType.selector);
+        vm.prank(rando);
+        validator.submitFundingProposal(
+            criteriaValue,
+            optionsDescriptions,
+            optionsRecipients,
+            optionsAmounts,
+            description,
+            ProposalValidator.ProposalType.ProtocolOrGovernorUpgrade
```
Comment (0xOneTony) [inline packages/contracts-bedrock/test/governance/ProposalValidator.t.sol]: again i would prefer if we fuzzed here the proposal type through the non-accepted values

**Review by 0xOneTony — CHANGES_REQUESTED** [review]
Tests need some further improvement, also added a comment about the `for` loops on the logic side

File: packages/contracts-bedrock/src/governance/ProposalValidator.sol (hunk 9 above)
Comment (0xChin) [reply packages/contracts-bedrock/src/governance/ProposalValidator.sol:308]: [perf: optimize for loops usage](https://github.com/defi-wonderland/optimism/commit/fc64dd0d4)

File: packages/contracts-bedrock/test/governance/ProposalValidator.t.sol (hunk 13 above)
Comment (0xChin) [reply packages/contracts-bedrock/test/governance/ProposalValidator.t.sol]: [refactor(test): improve tests legibility](https://github.com/defi-wonderland/optimism/commit/fc0135e64)

File: packages/contracts-bedrock/test/governance/ProposalValidator.t.sol (hunk 14 above)
Comment (0xChin) [reply packages/contracts-bedrock/test/governance/ProposalValidator.t.sol]: [test: use fuzzing instead of individual tests for edge cases](https://github.com/defi-wonderland/optimism/commit/75657d5b6)

File: packages/contracts-bedrock/test/governance/ProposalValidator.t.sol (hunk 16 above)
Comment (0xChin) [reply packages/contracts-bedrock/test/governance/ProposalValidator.t.sol]: [test: use fuzzing instead of individual tests for edge cases](https://github.com/defi-wonderland/optimism/commit/75657d5b6)

File: packages/contracts-bedrock/test/governance/ProposalValidator.t.sol (hunk 18 above)
Comment (0xChin) [reply packages/contracts-bedrock/test/governance/ProposalValidator.t.sol:848]: [test: use fuzzing instead of individual tests for edge cases](https://github.com/defi-wonderland/optimism/commit/75657d5b6)

File: packages/contracts-bedrock/test/governance/ProposalValidator.t.sol (hunk 11 above)
Comment (0xChin) [reply packages/contracts-bedrock/test/governance/ProposalValidator.t.sol]: [test: improve submitFundingProposal assertions](https://github.com/defi-wonderland/optimism/commit/4d94daf2d)

File: packages/contracts-bedrock/test/governance/ProposalValidator.t.sol (hunk 12 above)
Comment (0xChin) [reply packages/contracts-bedrock/test/governance/ProposalValidator.t.sol]: [test: use fuzzing instead of individual tests for edge cases](https://github.com/defi-wonderland/optimism/commit/75657d5b6)

File: packages/contracts-bedrock/test/governance/ProposalValidator.t.sol (hunk 15 above)
Comment (0xChin) [reply packages/contracts-bedrock/test/governance/ProposalValidator.t.sol]: [test: use fuzzing instead of individual tests for edge cases](https://github.com/defi-wonderland/optimism/commit/75657d5b6)

File: packages/contracts-bedrock/test/governance/ProposalValidator.t.sol (hunk 10 above)
Comment (0xChin) [reply packages/contracts-bedrock/test/governance/ProposalValidator.t.sol]: [test: fuzz proposer](https://github.com/defi-wonderland/optimism/commit/d81faa97e)

File: packages/contracts-bedrock/test/governance/ProposalValidator.t.sol (hunk 17 above)
Comment (0xChin) [reply packages/contracts-bedrock/test/governance/ProposalValidator.t.sol]: [chore(test): remove unused setup variables](https://github.com/defi-wonderland/optimism/commit/878556cf0)

Comment (0xChin) [reply packages/contracts-bedrock/test/governance/ProposalValidator.t.sol]: given that these values will mostly be fuzzed and that the values can vary depending on the proposal that is being made, for example funding proposals require being given amount, while council member election proposal don't, I'm not sure having them in the `Init` is the way to go. wdyt?

File: packages/contracts-bedrock/test/governance/ProposalValidator.t.sol (hunk 18 above)
Comment (0xChin) [reply packages/contracts-bedrock/test/governance/ProposalValidator.t.sol]: [test: use fuzzing for sad submitFundingProposal paths](https://github.com/defi-wonderland/optimism/commit/e7b0e6339)

File: packages/contracts-bedrock/test/governance/ProposalValidator.t.sol (hunk 17 above)
Comment (0xOneTony) [reply packages/contracts-bedrock/test/governance/ProposalValidator.t.sol]: I think with the usage of `_setGovernanceFundProposalType` its enough, the rest of the hardcoded values should be fuzzed where possible

File: packages/contracts-bedrock/test/governance/ProposalValidator.t.sol [hunk 19]
```diff
@@ -110,6 +125,50 @@ contract ProposalValidator_Init is CommonTest {
...
+    /// @notice Helper function to set CouncilBudget proposal type data.
+    function _setCouncilBudgetProposalType() internal {
+        _setProposalTypeData(
+            ProposalValidator.ProposalType.CouncilBudget,
+            ProposalValidator.ProposalTypeData({
+                requiredApprovals: PROPOSAL_REQUIRED_APPROVALS,
+                proposalVotingModule: 4
+            })
+        );
+    }
```
Comment (0xOneTony) [inline packages/contracts-bedrock/test/governance/ProposalValidator.t.sol:164]: Why is `proposalVotingModule` in one type `3` and on the other type `4` ?

File: packages/contracts-bedrock/test/governance/ProposalValidator.t.sol [hunk 20]
```diff
@@ -578,11 +714,356 @@ contract ProposalValidator_HashProposalWithModule_Test is ProposalValidator_Init
...
+
+        vm.expectRevert(ProposalValidator.ProposalValidator_ProposalTypesDataLengthMismatch.selector);
+        vm.prank(rando);
+        validator.submitFundingProposal(
+            criteriaValue,
+            mismatchedDescriptions,
+            matchingRecipients,
+            matchingAmounts,
+            description,
+            ProposalValidator.ProposalType.GovernanceFund
```
Comment (0xOneTony) [inline packages/contracts-bedrock/test/governance/ProposalValidator.t.sol:876]: Why dont we fuzz the proposal type? to make sure it fails in both cases

File: packages/contracts-bedrock/test/governance/ProposalValidator.t.sol [hunk 21]
```diff
@@ -578,11 +714,356 @@ contract ProposalValidator_HashProposalWithModule_Test is ProposalValidator_Init
...
+
+        vm.expectRevert(ProposalValidator.ProposalValidator_ProposalTypesDataLengthMismatch.selector);
+        vm.prank(rando);
+        validator.submitFundingProposal(
+            criteriaValue,
+            matchingDescriptions,
+            mismatchedRecipients,
+            matchingAmounts,
+            description,
+            ProposalValidator.ProposalType.GovernanceFund
```
Comment (0xOneTony) [inline packages/contracts-bedrock/test/governance/ProposalValidator.t.sol:904]: Same with this and the rest of the tests that only use `GovernanceFund` instead of fuzzing between the 2 types

File: packages/contracts-bedrock/test/governance/ProposalValidator.t.sol [hunk 22]
```diff
@@ -578,11 +714,356 @@ contract ProposalValidator_HashProposalWithModule_Test is ProposalValidator_Init
...
+
+        optionsRecipients = new address[](2);
+        optionsRecipients[0] = makeAddr("recipient1");
+        optionsRecipients[1] = makeAddr("recipient2");
+
+        optionsAmounts = new uint256[](2);
+        optionsAmounts[0] = 1000 ether;
+        optionsAmounts[1] = 500 ether;
+
+        description = "Test funding proposal";
```
Comment (0xOneTony) [inline packages/contracts-bedrock/test/governance/ProposalValidator.t.sol:836]: still not a fan of this many values being hardcoded, can we improve this?

**Review by 0xOneTony — CHANGES_REQUESTED** [review]

File: packages/contracts-bedrock/test/governance/ProposalValidator.t.sol [hunk 23]
```diff
@@ -578,11 +714,356 @@ contract ProposalValidator_HashProposalWithModule_Test is ProposalValidator_Init
...
+    }
+
+    function testFuzz_submitFundingProposal_invalidProposalType_reverts(uint8 proposalTypeValue) public {
+        // Bound to proposal types that are NOT funding proposals (0, 1, 2)
+        // Valid funding proposal types are GovernanceFund (3) and CouncilBudget (4)
+        proposalTypeValue = uint8(bound(proposalTypeValue, 0, 2));
+        ProposalValidator.ProposalType proposalType = ProposalValidator.ProposalType(proposalTypeValue);
+
+        vm.expectRevert(ProposalValidator.ProposalValidator_InvalidFundingProposalType.selector);
+        vm.prank(rando);
```
Comment (0xOneTony) [inline packages/contracts-bedrock/test/governance/ProposalValidator.t.sol:846]: We should totally remove this `rando` variable or improve it. The issue is the naming of the variable is wrong since this is not a random address but a hardcoded one that is the same every time. I think we should remove this and always fuzz the proposer or use the accepted address.

File: packages/contracts-bedrock/test/governance/ProposalValidator.t.sol (hunk 19 above)
Comment (0xChin) [reply packages/contracts-bedrock/test/governance/ProposalValidator.t.sol:164]: you're absolutely right [refactor: normalize funding proposal voting modules in global variable](https://github.com/defi-wonderland/optimism/commit/5ede7dd81)

File: packages/contracts-bedrock/test/governance/ProposalValidator.t.sol (hunk 20 above)
Comment (0xChin) [reply packages/contracts-bedrock/test/governance/ProposalValidator.t.sol:876]: [test: fuzz proposal types for revert cases](https://github.com/defi-wonderland/optimism/commit/03dc4d8af)

File: packages/contracts-bedrock/test/governance/ProposalValidator.t.sol (hunk 21 above)
Comment (0xChin) [reply packages/contracts-bedrock/test/governance/ProposalValidator.t.sol:904]: [test: fuzz proposal types for revert cases](https://github.com/defi-wonderland/optimism/commit/03dc4d8af)

File: packages/contracts-bedrock/test/governance/ProposalValidator.t.sol (hunk 22 above)
Comment (0xChin) [reply packages/contracts-bedrock/test/governance/ProposalValidator.t.sol:836]: [refactor: use minimal values for funding proposal revert cases tests](https://github.com/defi-wonderland/optimism/commit/e2df8d63a)

File: packages/contracts-bedrock/test/governance/ProposalValidator.t.sol (hunk 23 above)
Comment (0xChin) [reply packages/contracts-bedrock/test/governance/ProposalValidator.t.sol:846]: [refactor: use more descriptive variables for testing](https://github.com/defi-wonderland/optimism/commit/7745ca4e9)

**Review by 0xOneTony — APPROVED** [review]
lgtm! 🔥
